File: problems/problemset/expression-evaluator/Python/solution/evaluator.py
import math

class Evaluator:
    def __init__(self):
        self._vars: dict[str, float] = {}

    def set(self, name: str, value: float) -> None:
        self._vars[name] = value

    def eval(self, expression: str) -> float:
        self._src = expression
        self._pos = 0
        return self._expr()

    def _skip(self):
        while self._pos < len(self._src) and self._src[self._pos].isspace():
            self._pos += 1

    def _expr(self):
        v = self._term()
        self._skip()
        while self._pos < len(self._src) and self._src[self._pos] in '+-':
            op = self._src[self._pos]; self._pos += 1
            r = self._term()
            v = v + r if op == '+' else v - r
            self._skip()
        return v

    def _term(self):
        v = self._unary()
        self._skip()
        while self._pos < len(self._src) and self._src[self._pos] in '*/':
            op = self._src[self._pos]; self._pos += 1
            r = self._unary()
            v = v * r if op == '*' else v / r
            self._skip()
        return v

    def _unary(self):
        self._skip()
        if self._pos < len(self._src) and self._src[self._pos] == '-':
            self._pos += 1
            return -self._unary()
        return self._primary()

    def _primary(self):
        self._skip()
        s = self._src
        p = self._pos
        if p < len(s) and s[p] == '(':
            self._pos += 1
            v = self._expr()
            self._skip()
            if self._pos < len(s) and s[self._pos] == ')':
                self._pos += 1
            return v
        if p < len(s) and (s[p].isdigit() or s[p] == '.'):
            start = p
            while self._pos < len(s) and (s[self._pos].isdigit() or s[self._pos] == '.'):
                self._pos += 1
            return float(s[start:self._pos])
        if p < len(s) and s[p].isalpha():
            start = p
            while self._pos < len(s) and s[self._pos].isalnum():
                self._pos += 1
            name = s[start:self._pos]
            self._skip()
            if self._pos < len(s) and s[self._pos] == '(':
                self._pos += 1
                args = []
                self._skip()
                if self._pos < len(s) and s[self._pos] != ')':
                    args.append(self._expr())
                    self._skip()
                    while self._pos < len(s) and s[self._pos] == ',':
                        self._pos += 1
                        args.append(self._expr())
                        self._skip()
                if self._pos < len(s) and s[self._pos] == ')':
                    self._pos += 1
                return {'min': min, 'max': max, 'abs': abs, 'pow': math.pow}.get(name, lambda *a: 0)(*args)
            return self._vars.get(name, 0.0)
        return 0.0

File: problems/problemset/expression-evaluator/Rust/solution/evaluator.rs
use std::collections::HashMap;

pub struct Evaluator {
    vars: HashMap<String, f64>,
}

impl Evaluator {
    pub fn new() -> Self {
        Self { vars: HashMap::new() }
    }

    pub fn set(&mut self, name: &str, value: f64) {
        self.vars.insert(name.to_string(), value);
    }

    pub fn eval(&mut self, expression: &str) -> f64 {
        let mut p = Parser { src: expression.as_bytes(), pos: 0, vars: &self.vars };
        p.expr()
    }
}

struct Parser<'a> {
    src: &'a [u8],
    pos: usize,
    vars: &'a HashMap<String, f64>,
}

impl<'a> Parser<'a> {
    fn skip(&mut self) {
        while self.pos < self.src.len() && self.src[self.pos].is_ascii_whitespace() {
            self.pos += 1;
        }
    }

    fn peek(&self) -> Option<u8> {
        self.src.get(self.pos).copied()
    }

    fn expr(&mut self) -> f64 {
        let mut v = self.term();
        self.skip();
        while matches!(self.peek(), Some(b'+') | Some(b'-')) {
            let op = self.src[self.pos]; self.pos += 1;
            let r = self.term();
            v = if op == b'+' { v + r } else { v - r };
            self.skip();
        }
        v
    }

    fn term(&mut self) -> f64 {
        let mut v = self.unary();
        self.skip();
        while matches!(self.peek(), Some(b'*') | Some(b'/')) {
            let op = self.src[self.pos]; self.pos += 1;
            let r = self.unary();
            v = if op == b'*' { v * r } else { v / r };
            self.skip();
        }
        v
    }

    fn unary(&mut self) -> f64 {
        self.skip();
        if self.peek() == Some(b'-') { self.pos += 1; return -self.unary(); }
        self.primary()
    }

    fn primary(&mut self) -> f64 {
        self.skip();
        match self.peek() {
            Some(b'(') => {
                self.pos += 1;
                let v = self.expr();
                self.skip();
                if self.peek() == Some(b')') { self.pos += 1; }
                v
            }
            Some(c) if c.is_ascii_digit() || c == b'.' => {
                let start = self.pos;
                while matches!(self.peek(), Some(c) if c.is_ascii_digit() || c == b'.') {
                    self.pos += 1;
                }
                std::str::from_utf8(&self.src[start..self.pos]).unwrap().parse().unwrap_or(0.0)
            }
            Some(c) if c.is_ascii_alphabetic() => {
                let start = self.pos;
                while matches!(self.peek(), Some(c) if c.is_ascii_alphanumeric()) {
                    self.pos += 1;
                }
                let name = std::str::from_utf8(&self.src[start..self.pos]).unwrap();
                self.skip();
                if self.peek() == Some(b'(') {
                    self.pos += 1;
                    let mut args = vec![];
                    self.skip();
                    if self.peek() != Some(b')') {
                        args.push(self.expr());
                        self.skip();
                        while self.peek() == Some(b',') {
                            self.pos += 1;
                            args.push(self.expr());
                            self.skip();
                        }
                    }
                    if self.peek() == Some(b')') { self.pos += 1; }
                    return match name {
                        "min" => args[0].min(args[1]),
                        "max" => args[0].max(args[1]),
                        "abs" => args[0].abs(),
                        "pow" => args[0].powf(args[1]),
                        _ => 0.0,
                    };
                }
                *self.vars.get(name).unwrap_or(&0.0)
            }
            _ => 0.0,
        }
    }
}
